Extensions/Today/TodayViewController.swift

```diff
@@ -9,123 +9,169 @@ import SnapKit
+            make.centerX.equalTo(view.snp_centerX)
+            make.height.equalTo(buttonContainer.snp_height).multipliedBy(TodayUX.copiedLinkHeightOfButtonMultple)
         }
 
         view.snp_makeConstraints { make in
-            make.height.equalTo(buttonContainer.snp_height).priorityLow()
+            let multiple = !hasCopiedURL ? 1.0 : (1.0 + TodayUX.copiedLinkHeightOfButtonMultple)
+            make.height.equalTo(self.buttonContainer.snp_height).multipliedBy(multiple)
         }
-        view.systemLayoutSizeFittingSize(CGSizeZero)
+    }
 
-        updateOpenClipboardButton()
+    func updateCopiedLink() {
+        if let url = self.copiedURL {
+            self.openCopiedLinkButton.hidden = false
+            self.openCopiedLinkButton.subtitleLabel.hidden = SystemUtils.isDeviceLocked()
+            self.openCopiedLinkButton.subtitleLabel.text = url.absoluteString
+        } else {
+            self.openCopiedLinkButton.hidden = true
+        }
+
+        self.view.setNeedsLayout()
     }
 
     func widgetMarginInsetsForProposedMarginInsets(defaultMarginInsets: UIEdgeInsets) -> UIEdgeInsets {
-        return UIEdgeInsetsMake(0, 0, 0, 0)
+        return UIEdgeInsetsMake(0, 0, TodayUX.verticalWidgetMargin, 0)
     }
 
     override func viewDidLayoutSubviews() {
         super.viewDidLayoutSubviews()
     }
 
-    private func alignButton(button: UIButton, withLabel label: UILabel, leftOf leftLabel: UILabel? = nil) {
-        label.snp_makeConstraints { make in
-            make.centerX.equalTo(button.snp_centerX)
-            make.centerY.equalTo(button.snp_centerY).offset(39)
-            if let leftLabel = leftLabel {
-                make.centerX.equalTo(leftLabel.snp_centerX).offset(44).priorityLow()
-            }
+    private func alignButton(leftButton: ImageButtonWithLabel, rightButton: ImageButtonWithLabel) {
+        rightButton.label.snp_makeConstraints { make in
+            make.centerY.equalTo(leftButton.label.snp_centerY)
+            make.left.equalTo(leftButton.label.snp_right).offset(44).priorityLow()
         }
     }
 
-    override func didReceiveMemoryWarning() {
-        super.didReceiveMemoryWarning()
-        // Dispose of any resources that can be recreated.
-    }
-    
     func widgetPerformUpdateWithCompletionHandler(completionHandler: ((NCUpdateResult) -> Void)) {
         // Perform any setup necessary in order to update the view.
-        updateOpenClipboardButton()
 
         // If an error is encountered, use NCUpdateResult.Failed
         // If there's no update required, use NCUpdateResult.NoData
         // If there's an update, use NCUpdateResult.NewData
-
+        dispatch_async(dispatch_get_main_queue()) {
+            // updates need to be made on the main thread
+            self.updateCopiedLink()
+            // and we need to call the completion handler in every branch.
+            completionHandler(NCUpdateResult.NewData)
+        }
         completionHandler(NCUpdateResult.NewData)
```

Comment:
this will call the completion handler twice. Is that right?

Also, are these comments still valid as we explicitly set the result as `NCUpdateResult.NewData` and never any of the other options?